apps/manager_app/card_training.c

```diff
@@ -0,0 +1,171 @@
+
+/**
+ * TODO: Replace with api provided by manager app
+ *
+ * @param error_code
+ */
+static void send_training_error(uint32_t error_code);
+
+/*****************************************************************************
+ * STATIC VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * GLOBAL VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+
+static void send_message_to_host(manager_train_card_response_t *training) {
+  manager_result_t result =
+      get_manager_result_template(MANAGER_RESULT_TRAIN_CARD_TAG);
+  uint8_t result_buffer[1024] = {0};
+
+  memcpy(&result.train_card, training, sizeof(manager_train_card_response_t));
+  ASSERT(encode_and_send_manager_result(
+      &result, result_buffer, sizeof(result_buffer)));
+}
+
+static void send_training_error(uint32_t error_code) {
+  manager_train_card_response_t training =
+      MANAGER_TRAIN_CARD_RESPONSE_INIT_ZERO;
+  training.which_response = MANAGER_TRAIN_CARD_RESPONSE_COMMON_ERROR_TAG;
+  training.common_error.which_error = ERROR_COMMON_ERROR_UNKNOWN_ERROR_TAG;
+  training.common_error.unknown_error = error_code;
+  send_message_to_host(&training);
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+
+void manager_card_training(manager_query_t *query) {
+  LOG_SWV("%s (%d)\n", __func__, __LINE__);
+  if (NULL == query) {
+    // TODO: verify card training query tag for completeness
+    return;
+  }
+  // TODO: verify on-boarding state for safety
+  core_status_set_device_waiting_on(CORE_DEVICE_WAITING_ON_BUSY_IP_CARD);
+
+  instruction_scr_init("Place any card below the device to test card tapping",
+                       NULL);
+
+  check_pairing_result_t pair_result = {false, 0};
+  manager_train_card_response_t result = MANAGER_TRAIN_CARD_RESPONSE_INIT_ZERO;
+  result.which_response = MANAGER_TRAIN_CARD_RESPONSE_RESULT_TAG;
+  card_error_type_e status = card_check_pairing(&pair_result);
+  if (CARD_OPERATION_SUCCESS != status) {
+    LOG_SWV("%s (%d)\n", __func__, __LINE__);
+    send_training_error(1);
+    return;
+  }
+  result.result.card_paired = pair_result.is_paired;
+
+  // always pair the card
+  // TODO: Update the pairing flow to support flexible storage of shared key
+
+  // TODO: Fetch wallet list; send dummy data
+  result.result.card_paired = false;
+  result.result.wallet_list_count = 0;
+  send_message_to_host(&result);
+
+  char msg[64] = "";
+  snprintf(msg, sizeof(msg), "Card #%d Tapped", pair_result.card_number);
+  delay_scr_init(msg, DELAY_TIME);
```

Comment:
How will we synchronise this delay with Cysync?

Reply:
mutual understanding between them

Reply:
mutual understanding between them